x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
-            }
-
-            @Override
-            public int hashCode() {
-                return pattern.hashCode();
+        @Override
+        public Query wildcardQuery(String wildcardPattern, RewriteMethod method, QueryShardContext context) {
+            
+            Automaton dvAutomaton = WildcardQuery.toAutomaton(new Term(name(), wildcardPattern));
+            
+            String ngramIndexPattern = toLowerCase(wildcardPattern);
+            // Parse the wildcard, marking ? and * chars with a single invalid char
+            // to create a simple finite automaton we can parse
+            List<Automaton> automata = new ArrayList<>();            
+            for (int i = 0; i < ngramIndexPattern.length();) {
+              final int c = ngramIndexPattern.codePointAt(i);
+              int length = Character.charCount(c);
+              switch(c) {
+                case WildcardQuery.WILDCARD_STRING: 
+                  automata.add(Automata.makeChar(INVALID_AUTOMATON_CHAR));
+                  break;
+                case WildcardQuery.WILDCARD_CHAR:
+                  automata.add(Automata.makeChar(INVALID_AUTOMATON_CHAR));
+                  break;
+                case WildcardQuery.WILDCARD_ESCAPE:
+                  // add the next codepoint instead, if it exists
+                  if (i + length < wildcardPattern.length()) {
+                    final int nextChar = wildcardPattern.codePointAt(i + length);
+                    length += Character.charCount(nextChar);
+                    automata.add(Automata.makeChar(nextChar));
+                  } else {
+                      automata.add(Automata.makeChar(c));                      
+                  }
+                  break;
+                      
+                default:
+                  automata.add(Automata.makeChar(c));
+              }
+              i += length;
             }
+            
+            Automaton approxAutomata = Operations.concatenate(automata);                
 
-            @Override
-            public boolean equals(Object obj) {
-                PatternStructure other = (PatternStructure) obj;
-                return pattern.equals(other.pattern);
+            return createApproxAndVerifyQuery(approxAutomata, dvAutomaton, wildcardPattern);
+        }
+        
+        @Override
+        public Query regexpQuery(String value, int flags, int maxDeterminizedStates, RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[regexp] queries cannot be executed when '" + ALLOW_EXPENSIVE_QUERIES.getKey() + "' is set to false."
+                );
             }
+            failIfNotIndexed();
+            RegExp regex = new RegExp(value, flags);
+            Automaton automaton = regex.toAutomaton(maxDeterminizedStates);
+            ApproximateRegExp ngramRegex = new ApproximateRegExp(toLowerCase(value), flags);
```

Comment:
I don't think we should lowercase eagerly here. Lowercasing can happen in the `ApproximateRegExp` directly but I wonder if this should be tackled in a follow up ?